src/AutoRest.CSharp/Common/Output/Models/Types/ModelFactoryTypeProvider.cs

```diff
@@ -31,30 +31,19 @@ internal sealed class ModelFactoryTypeProvider : TypeProvider
         protected override string DefaultName { get; }
         protected override string DefaultAccessibility { get; }
 
-        // TODO: remove this intermediate state once we generate it before output types
         private IReadOnlyList<Method>? _methods;
+        public IReadOnlyList<Method> Methods => _methods ??= BuildMethods().ToArray();
 
-        // This method should only be called from OutputMethods as intermediate state.
-        private IReadOnlyList<Method> ShouldNotBeUsedForOutput([CallerMemberName] string caller = "")
+        private IEnumerable<Method> BuildMethods()
         {
-            Debug.Assert(caller == nameof(Methods) || caller == nameof(SignatureType), $"This method should not be used for output. Caller: {caller}");
-            return _methods ??= _models.Select(CreateMethod).ToList();
-        }
-
-        // TODO: remove this intermediate state once we generate it before output types
```

Comment:
So, it seems like we would never be able to move this before constructing the output types because we need to have the generated method signatures to get the changeset combined with previous contract and custom potion.

But, is it possible to construct signatures beforehand and move this breaking change resolving functionality ahead? And then, we just need to input the prepared methods & properties into the output types instead of doing them here. That's the final vision from @m-nash. 

Again, that's why I would like to put this logic in a central place to be easily moved around later.